bigbluebutton/greenlight
# GLv3-3202-frontend-structure

Frontend:
	          - Set a new project structure for the front end JavaScript code influenced by the
	          react-router-demo https://stackblitz.com/edit/github-agqlf5/ :
	              -> main.jsx: An entry point where Routes, Layouts, Global context is provided.
	              -> app.jsx: The parent route component for the `/` endpoint.
	              -> components/: A directory where custom components resides.
	              -> components/styling: A directory where custom components with no business logic resides (i.e. custom buttons).
	              -> components/forms: A directory where custom forms and form related components resides.
	              -> routes: A directory where sub routes resides.
	          - Created a `comonents/styling/buttons/ButtonLink.jsx` component that wraps/connects a react bootstrap button component with the react router API.

                  TODO:
  		    - Create the signup form.
## Description

Defined a project structure for GLv3 frontend code.
## Testing Steps

1. Pull the code.
2. Install the dependencies `bundle install && npm|yarn install`.
3. Clean the previous assets build by running `rm app/assets/builds/*` (This won't remove .keep since it's hidden).
4. Clean the database and tmp files for a better isolation by running `rails tmp:clear && rails db:schema:cache:clear && rails db:migrate:reset`
5. Run the linter and specs `bundle exec rubocop --parallel && bundle exec rspec`
6. Run `./bin/dev` to run the assets builders processes and the Puma server all at once.

To test the API:
    1. Make a POST request to `/api/v1/users.json`.
     1.1 The request has to have an `Accept: application/json` header.
     1.2. The request body should have a JSON representation with this format: 
     `{ user: { :name, :email, :password, :password_confirmation } }` .
     -> When not specifying the `.json` format the API should render back the `components#index` so that CSR handles the request.
     -> When not specifying the `Accept` header correctly the same scenario should occur.
     -> For unknown routes in the server side the CSR should take charge of handling it.
     -> Not respecting the user data required format will return a 400 BAD request with a JSON object in the response body having an errors array with a message **Invalid or Missing parameters.**
     -> Any unpermitted values given in the user data will be filtered, only `name, email ,password,...` will pass.
 
 To test the UI:
    1.  Entering `/` should render this:    
![image](https://user-images.githubusercontent.com/29759616/159480004-b75dc0b6-2d4e-480b-a921-525e657330d2.png)

    2. Entering `/signup` should render this:    
![image](https://user-images.githubusercontent.com/29759616/159480060-0ad38785-4c50-4bd0-a4c2-ca76a51a32b5.png)

    3. Entering anything else should render this:
![image](https://user-images.githubusercontent.com/29759616/159480130-d2c46c44-f3f7-42b2-83f8-dd821e9f253b.png)

    4:  Clicking on the Signup button should have the same outcome as 2). 
## Screenshots (if appropriate):
